File: AndroidJettyContainer/app/src/main/java/com/ipostu/androidjettycontainer/MainActivity.java
package com.ipostu.androidjettycontainer;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.text.Html;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ScrollView;
import android.widget.TextView;

import com.ipostu.androidjettycontainer.core.log.AndroidLog;
import com.ipostu.androidjettycontainer.util.AndroidInfo;
import com.ipostu.androidjettycontainer.util.AppTools;

import java.util.Date;

public class MainActivity extends PermissionActivity {

    private static final String TAG = "Jetty";

    public static final String __START_ACTION = "com.ipostu.androidjettycontainer.start";
    public static final String __STOP_ACTION = "com.ipostu.androidjettycontainer.stop";

    public static final String __PORT = "com.ipostu.androidjettycontainer.port";
    public static final String __NIO = "com.ipostu.androidjettycontainer.nio";
    public static final String __SSL = "com.ipostu.androidjettycontainer.ssl";

    public static final String __CONSOLE_PWD = "com.ipostu.androidjettycontainer.console";
    public static final String __PORT_DEFAULT = "8080";
    public static final boolean __NIO_DEFAULT = true;
    public static final boolean __SSL_DEFAULT = false;

    public static final String __CONSOLE_PWD_DEFAULT = "admin";

    public static final int __SETUP_PROGRESS_DIALOG = 0;
    public static final int __SETUP_DONE = 2;
    public static final int __SETUP_RUNNING = 1;
    public static final int __SETUP_NOTDONE = 0;

    private Button startButton;
    private Button stopButton;
    private Button configButton;
    private TextView footer;
    private TextView info;
    private TextView console;
    private ScrollView consoleScroller;
    private StringBuilder consoleBuffer = new StringBuilder();
    private Runnable scrollTask;
    private ProgressDialog progressDialog;
    private Thread progressThread;
    private Handler handler;
    private BroadcastReceiver bcastReceiver;

    static {
        // Ensure parsing is not validating - does not work with android
        System.setProperty("org.eclipse.jetty.xml.XmlParser.Validating", "false");

        // Bridge Jetty logging to Android logging
        System.setProperty("org.eclipse.jetty.util.log.class", "com.ipostu.androidjettycontainer.core.log.AndroidLog");
        org.eclipse.jetty.util.log.Log.setLog(new AndroidLog());
    }

    class ConsoleScrollTask implements Runnable {
        public void run() {
            consoleScroller.fullScroll(View.FOCUS_DOWN);
        }
    }

    public MainActivity() {
        super();
        handler = new Handler() {
            public void handleMessage(Message msg) {
                int total = msg.getData().getInt("prog");
                progressDialog.setProgress(total);
                if (total >= 100) {
                    dismissDialog(__SETUP_PROGRESS_DIALOG);
                }
            }

        };
    }

    @Override
    protected void onPermissionDenied() {
    }

    @Override
    protected void onPermissionGranted() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        startButton = (Button) findViewById(R.id.start);
        stopButton = (Button) findViewById(R.id.stop);
        configButton = (Button) findViewById(R.id.config);
        final Button downloadButton = (Button) findViewById(R.id.download);

        IntentFilter filter = new IntentFilter();
        filter.addAction(__START_ACTION);
        filter.addAction(__STOP_ACTION);
        filter.addCategory("default");

        bcastReceiver =
                new BroadcastReceiver() {

                    public void onReceive(Context context, Intent intent) {
                        if (__START_ACTION.equalsIgnoreCase(intent.getAction())) {
                            startButton.setEnabled(false);
                            configButton.setEnabled(false);
                            stopButton.setEnabled(true);
                            consolePrint("<br/>Started Jetty at %s", new Date());
                            String[] connectors = intent.getExtras().getStringArray("connectors");
                            if (null != connectors) {
                                for (int i = 0; i < connectors.length; i++)
                                    consolePrint(connectors[i]);
                            }

                            AppTools.printNetworkInterfaces(consoleBuffer);

                            if (AndroidInfo.isOnEmulator(MainActivity.this)) {
                                consolePrint("Set up port forwarding to see i-jetty outside of the emulator.");
                            } else {
                                consolePrint("");
                            }
                        } else if (__STOP_ACTION.equalsIgnoreCase(intent.getAction())) {
                            startButton.setEnabled(true);
                            configButton.setEnabled(true);
                            stopButton.setEnabled(false);
                            consolePrint("<br/> Jetty stopped at %s", new Date());
                        }
                    }

                };

        registerReceiver(bcastReceiver, filter);


        // Watch for button clicks.
        startButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                if (ProgressThread.isUpdateNeeded(MainActivity.this))
                    AppTools.showQuickToast(MainActivity.this, R.string.loading);
                else {
                    //TODO get these values from editable UI elements
                    Intent intent = new Intent(MainActivity.this, JettyServerService.class);
                    intent.putExtra(__PORT, __PORT_DEFAULT);
                    intent.putExtra(__NIO, __NIO_DEFAULT);
                    intent.putExtra(__SSL, __SSL_DEFAULT);
                    intent.putExtra(__CONSOLE_PWD, __CONSOLE_PWD_DEFAULT);
                    startService(intent);
                }
            }
        });

        stopButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                stopService(new Intent(MainActivity.this, JettyServerService.class));
            }
        });


        configButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                EditorPreferenceActivity.show(MainActivity.this);
            }
        });


        downloadButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                DownloaderActivity.show(MainActivity.this);
            }
        });

        info = (TextView) findViewById(R.id.info);
        footer = (TextView) findViewById(R.id.footer);
        console = (TextView) findViewById(R.id.console);
        consoleScroller = (ScrollView) findViewById(R.id.consoleScroller);

        StringBuilder infoBuffer = new StringBuilder();
        try {
            PackageInfo pi = getPackageManager().getPackageInfo(getPackageName(), 0);
            infoBuffer.append(AppTools.formatJettyInfoLine("i-jetty version %s (%s)", pi.versionName, pi.versionCode));
        } catch (PackageManager.NameNotFoundException e) {
            infoBuffer.append(AppTools.formatJettyInfoLine("i-jetty version unknown"));
        }
        infoBuffer.append(AppTools.formatJettyInfoLine("On %s using Android version %s", AndroidInfo.getDeviceModel(), AndroidInfo.getOSVersion()));
        info.setText(Html.fromHtml(infoBuffer.toString()));


        StringBuilder footerBuffer = new StringBuilder();
        footerBuffer.append("<b>Project:</b> <a href=\"https://github.com/</a> <br/>");
        footerBuffer.append("<b>Server:</b> http://www.eclipse.org/jetty/ <br/>");
        footer.setText(Html.fromHtml(footerBuffer.toString()));
    }

    public void consolePrint(String format, Object... args) {
        String msg = String.format(format, args);
        if (msg.length() > 0) {
            consoleBuffer.append(msg).append("<br/>");
            console.setText(Html.fromHtml(consoleBuffer.toString()));
            Log.i(TAG, msg); // Only interested in non-empty lines being output to Log
        } else {
            consoleBuffer.append(msg).append("<br/>");
            console.setText(Html.fromHtml(consoleBuffer.toString()));
        }

        if (scrollTask == null) {
            scrollTask = new ConsoleScrollTask();
        }

        consoleScroller.post(scrollTask);
    }

    @Override
    protected void onDestroy() {
        if (bcastReceiver != null)
            unregisterReceiver(bcastReceiver);
        super.onDestroy();
    }

    @Override
    protected void onResume() {
        if (!SdCardUnavailableActivity.isExternalStorageAvailable()) {
            SdCardUnavailableActivity.show(this);
        } else {
            //work out if we need to do the installation finish step
            //or not. We do it iff:
            // - there is no previous jetty version on disk
            // - the previous version does not match the current version
            // - we're not already doing the update
            if (ProgressThread.isUpdateNeeded(MainActivity.this)) {
                setupJetty();
            }
        }

        if (JettyServerService.isRunning()) {
            startButton.setEnabled(false);
            configButton.setEnabled(false);
            stopButton.setEnabled(true);
        } else {
            startButton.setEnabled(true);
            configButton.setEnabled(true);
            stopButton.setEnabled(false);
        }
        super.onResume();
    }

    @Override
    protected Dialog onCreateDialog(int id) {
        switch (id) {
            case __SETUP_PROGRESS_DIALOG: {
                progressDialog = new ProgressDialog(MainActivity.this);
                progressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
                progressDialog.setMessage("Finishing initial install ...");

                return progressDialog;
            }
            default:
                return null;
        }
    }

    public void setupJetty() {
        showDialog(__SETUP_PROGRESS_DIALOG);
        progressThread = new ProgressThread(MainActivity.this, handler);
        progressThread.start();
    }

    public static void show(Context context) {
        final Intent intent = new Intent(context, MainActivity.class);
        context.startActivity(intent);
    }

}

File: app/src/main/java/com/ipostu/androidjettycontainer/MainActivity.kt
package com.ipostu.androidjettycontainer

import android.app.Dialog
import android.app.ProgressDialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.text.Html
import android.view.View
import android.widget.Button
import android.widget.ScrollView
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import org.eclipse.jetty.util.log.Log
import java.util.Date

class MainActivity : PermissionActivity() {

    companion object {

        const val __START_ACTION: String = "org.ijetty.server.app.start"
        const val __STOP_ACTION: String = "org.ijetty.server.app.stop"


        const val __PORT: String = "org.ijetty.server.app.port"
        const val __NIO: String = "org.ijetty.server.app.nio"
        const val __SSL: String = "org.ijetty.server.app.ssl"

        const val __CONSOLE_PWD: String = "org.ijetty.server.app.console"
        const val __PORT_DEFAULT: String = "8080"
        const val __NIO_DEFAULT: Boolean = true
        const val __SSL_DEFAULT: Boolean = false

        const val __CONSOLE_PWD_DEFAULT: String = "admin"

        const val __SETUP_PROGRESS_DIALOG: Int = 0

        // Static initializer block
        init {
            // Ensure parsing is not validating - does not work with android
            System.setProperty("org.eclipse.jetty.xml.XmlParser.Validating", "false")


            // Bridge Jetty logging to Android logging
            System.setProperty(
                "org.eclipse.jetty.util.log.class", "com.ipostu.androidjettycontainer.AndroidLog"
            )
            Log.setLog(AndroidLog())
        }
    }

    private val TAG = "Jetty"

    private val consoleBuffer = StringBuilder()
    private val handler: Handler

    private lateinit var startButton: Button
    private lateinit var stopButton: Button
    private lateinit var configButton: Button
    private lateinit var downloadButton: Button
    private lateinit var progressDialog: ProgressDialog
    private lateinit var consoleScroller: ScrollView
    private lateinit var console: TextView
    private lateinit var info: TextView
    private lateinit var footer: TextView


    private lateinit var broadcastReceiver: BroadcastReceiver
    private var scrollTask: Runnable? = null


    class ConsoleScrollTask(private val consoleScroller: ScrollView) : Runnable {
        override fun run() {
            consoleScroller.fullScroll(View.FOCUS_DOWN)
        }
    }

    init {
        handler = object : Handler() {
            override fun handleMessage(msg: Message) {
                val total = msg.data.getInt("prog")
                progressDialog.setProgress(total)
                if (total >= 100) {
                    dismissDialog(__SETUP_PROGRESS_DIALOG)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        startButton = findViewById(R.id.start)
        stopButton = findViewById(R.id.stop)
        configButton = findViewById(R.id.config)
        downloadButton = findViewById(R.id.download)

        consoleScroller = findViewById(R.id.consoleScroller)
        console = findViewById(R.id.console)
        info = findViewById(R.id.info)
        footer = findViewById(R.id.footer)

        val filter = IntentFilter()
        filter.addAction(__START_ACTION)
        filter.addAction(__STOP_ACTION)
        filter.addCategory("default")

        broadcastReceiver =
            object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    if (__START_ACTION.equals(intent.action, ignoreCase = true)) {
                        startButton.isEnabled = false
                        configButton.isEnabled = false
                        stopButton.isEnabled = true
                        consolePrint("<br/>Started Jetty at %s", Date())
                        val connectors = intent.extras!!.getStringArray("connectors")
                        if (null != connectors) {
                            for (i in connectors.indices) consolePrint(connectors[i])
                        }

                        AppTools.printNetworkInterfaces(consoleBuffer)

                        if (AndroidInfo.isOnEmulator(this@MainActivity)) {
                            consolePrint("Set up port forwarding to see i-jetty outside of the emulator.")
                        } else {
                            consolePrint("")
                        }
                    } else if (__STOP_ACTION.equals(intent.action, ignoreCase = true)) {
                        startButton.isEnabled = true
                        configButton.isEnabled = true
                        stopButton.isEnabled = false
                        consolePrint("<br/> Jetty stopped at %s", Date())
                    }
                }
            }

        registerReceiver(broadcastReceiver, filter)

        startButton.setOnClickListener {
            if (ProgressThread.isUpdateNeeded(this@MainActivity)) AppTools.showQuickToast(
                this@MainActivity,
                R.string.loading
            )
            else {
                //TODO get these values from editable UI elements
                val intent = Intent(
                    this@MainActivity,
                    JettyServerService::class.java
                )
                intent.putExtra(__PORT, __PORT_DEFAULT)
                intent.putExtra(__NIO, __NIO_DEFAULT)
                intent.putExtra(__SSL, __SSL_DEFAULT)
                intent.putExtra(__CONSOLE_PWD, __CONSOLE_PWD_DEFAULT)
                startService(intent)
            }
        }

        stopButton.setOnClickListener {
            stopService(
                Intent(
                    this@MainActivity,
                    JettyServerService::class.java
                )
            )
        }

        val infoBuffer = java.lang.StringBuilder()
        try {
            val pi = packageManager.getPackageInfo(packageName, 0)
            infoBuffer.append(
                AppTools.formatJettyInfoLine(
                    "i-jetty version %s (%s)",
                    pi.versionName,
                    pi.versionCode
                )
            )
        } catch (e: PackageManager.NameNotFoundException) {
            infoBuffer.append(AppTools.formatJettyInfoLine("i-jetty version unknown"))
        }
        infoBuffer.append(
            AppTools.formatJettyInfoLine(
                "On %s using Android version %s",
                AndroidInfo.getDeviceModel(),
                AndroidInfo.getOSVersion()
            )
        )
        info.text = Html.fromHtml(infoBuffer.toString())

        val footerBuffer = java.lang.StringBuilder()
        footerBuffer.append("<b>GitHub:</b>" +
                " <a href=\"https://github.com/IvanPostu\">https://github.com/IvanPostu</a> <br/>")
        footerBuffer.append("<b>Server:</b> http://www.eclipse.org/jetty/ <br/>")
        footer.text = Html.fromHtml(footerBuffer.toString())
    }

    override fun onPermissionDenied() {
    }

    override fun onPermissionGranted() {
    }


    override fun onCreateDialog(id: Int): Dialog? {
        when (id) {
            __SETUP_PROGRESS_DIALOG -> {
                progressDialog = ProgressDialog(this@MainActivity)
                progressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL)
                progressDialog.setMessage("Finishing initial install ...")

                return progressDialog
            }

            else -> return null
        }
    }

    private fun consolePrint(format: String?, vararg args: Any?) {
        val msg = String.format(format!!, *args)
        if (msg.length > 0) {
            consoleBuffer.append(msg).append("<br/>")
            console.setText(Html.fromHtml(consoleBuffer.toString()))
            android.util.Log.i(
                TAG,
                msg
            ) // Only interested in non-empty lines being output to Log
        } else {
            consoleBuffer.append(msg).append("<br/>")
            console.setText(Html.fromHtml(consoleBuffer.toString()))
        }

        if (scrollTask == null) {
            scrollTask = ConsoleScrollTask(consoleScroller)
        }

        consoleScroller.post(scrollTask)
    }
}